File: secant_mt.cpp
// Phương pháp dây cung
// dựa theo công thức sai số mục tiêu
// | x_k - x* | <= | f(x_k) | / m1

#include <cmath>
#include <iomanip>
#include <iostream>

using namespace std;

double mypow(double x, int n) {
  double res = 1;
  while (n--) res *= x;
  return res;
}

double f(double x) { return mypow(x, 5) - 3 * mypow(x, 3) + 2 * x * x - x + 5;}
double df(double x) { return 5*mypow(x, 4) - 9*x*x + 4*x - 1;}
double d2f(double x) { return 20* mypow(x, 3) - 18*x + 4; }

void solve(double a, double b, double epsilon) {
  double d, x_old;

  if (f(a) * d2f(a) > 0) {
    d = a;
    x_old = b;
  } else {
    d = b;
    x_old = a;
  }

  double m1 = min(abs(df(a)), abs(df(b)));
  int k = 1;

  cout << fixed << setprecision(10);
  cout << "Diem Fourier d = " << d << "\n";
  cout << "Diem bat dau x0 = " << x_old << "\n";
  cout << "----------------------------------------------------" << "\n";
  cout << setw(3) << "k"
       << setw(15) << "x_k"
       << setw(15) << "x_k+1"
       << setw(15) << "Sai so\n";

  while (true) {
    double x_new = x_old - (f(x_old) * (x_old - d)) / (f(x_old) - f(d));
    double error = abs(f(x_new)) / m1;

    cout << setw(3) << k
         << setw(15) << x_old
         << setw(15) << x_new
         << setw(15) << error << "\n";

    if (error <= epsilon) {
      cout << "----------------------------------------------------\n";
      cout << "Nghiem xap xi tim duoc: " << x_new << "\n";
      break;
    }

    x_old = x_new;
    k++;

    if (k > 100) {
      cout << "Lap qua 100 lan!\n";
      break;
    }
  }
}

int main() {
  double a, b, epsilon;
  if (!(cin >> a >> b >> epsilon))
    return 0;
  solve(a, b, epsilon);
  return 0;
}

File: 1d/secant_mt.py
# Phuong phap day cung
# dua theo cong thuc sai so muc tieu
# | x_k - x* | <= | f(x_k) | / m1

import sympy as sp

_x = sp.symbols("x")

# CHINH HAM f(x) O DAY:
_f_expr = _x**5 - 17 # <- Ham f(x), bien x go thanh _x

_df_expr = sp.diff(_f_expr, _x)
_d2f_expr = sp.diff(_df_expr, _x)

f = sp.lambdify(_x, _f_expr, "math")
df = sp.lambdify(_x, _df_expr, "math")
d2f = sp.lambdify(_x, _d2f_expr, "math")


def solve(a, b, epsilon, max_iter=100):
    if f(a) * d2f(a) > 0:
        d, x_old = a, b
    else:
        d, x_old = b, a

    m1 = min(abs(df(a)), abs(df(b)))
    k = 0

    print(f"Diem Fourier d = {d:.10f}")
    print(f"Diem bat dau x0 = {x_old:.10f}")
    print("-" * 52)
    print(f"{'k':>3}{'x_k':>15}{'x_k+1':>15}{'Sai so':>15}")

    while True:
        x_new = x_old - (f(x_old) * (x_old - d)) / (f(x_old) - f(d))
        error = abs(f(x_new)) / m1

        print(f"{k:>3}{x_old:>15.10f}{x_new:>15.10f}{error:>15.10f}")

        if error <= epsilon:
            print("-" * 52)
            print(f"Nghiem xap xi tim duoc: {x_new:.10f}")
            break

        x_old = x_new
        k += 1

        if k > max_iter:
            print(f"Lap qua {max_iter} lan!")
            break


def main():
    default_max_iter = 100
    data = input("Nhap a, b, epsilon va (tuy chon) so lan lap toi da: ").split()
    a, b, epsilon = map(float, data[:3])

    if len(data) > 3:
        max_iter = int(data[3])
    else:
        try:
            extra = input(f"So lan lap toi da (Enter = {default_max_iter}): ").strip()
        except EOFError:
            extra = ""
        max_iter = int(extra) if extra else default_max_iter

    solve(a, b, epsilon, max_iter)


if __name__ == "__main__":
    main()
